### ISSUE TITLE
What is the correct GC generation for static string field in Net6?

### ISSUE THREAD

**Issue by Li7ye** [dotnet/runtime#68544]
I create a string object as static field as below:
```csharp
static class Statics
{
    public static string str = "123";
}
```
I call `GC.GetGeneration(Statics.str)` method and **get **generation 0****.

I try to use Windbg with SOS extension to get `Statics.str` information about GC generation. However, **Windbg displays GC generation of `Statics.str` is 4**.

```
0:000> !gcwhere 18818002E90h
Address            Gen   Heap   segment            begin              allocated           size
0000018818002E90   4      0     0000018818000000   0000018818001000   0000018818005420    0x0(0)
```
I notice that the address of `Statics.str` is in `Pinned object heap`.  It looks like  `Pinned object heap` is logical generation 4 in generation 2 .
```
Number of GC Heaps: 1
generation 0 starts at 0x0000018800001030
generation 1 starts at 0x0000018800001018
generation 2 starts at 0x0000018800001000

Pinned object heap starts at 0x0000018818001000
0000018818000000  0000018818001000  0000018818005420  0000018818012000  0x4420(17440)  0x11000(69632)
```

**Comment by huoyaoyuan** [comment]
Is string literal special? Can it live in constant area of string table?
I think it's better to check some other type too.

**Comment by Li7ye** [comment]
> Is string literal special? Can it live in constant area of string table? I think it's better to check some other type too.

Hi Huo Yao Yuan,

As you said, I define `Foo` class as below:
```csharp
class Foo
{
  public int Id;
  public Foo(int id)
  {
    Id = id;
  }
}
```
And define a 'Foo' instance as static field:
```csharp
static class Statics
{
    public static Foo Foo  = new(3);
}
```

Main method:
```csharp
var foo = Statics.Foo;
var generation = GC.GetGeneration(foo);
Console.WriteLine(generation);
Debugger.Break();
Console.ReadKey();
```

The result of `GC.GetGeneration(Statics.Foo)` still is 0 and `!gcwhere`  outputs GC generation is 4 as well for `Statics.Foo`.

```
D:\MyCodes\DebugTest\DebugTest\DebugTest\Program.cs @ 4:
00007ffa`5f802bf0 55              push    rbp
00007ffa`5f802bf1 4883ec40        sub     rsp,40h
00007ffa`5f802bf5 488d6c2440      lea     rbp,[rsp+40h]
00007ffa`5f802bfa 33c0            xor     eax,eax
00007ffa`5f802bfc 488945f8        mov     qword ptr [rbp-8],rax
00007ffa`5f802c00 8945f4          mov     dword ptr [rbp-0Ch],eax
00007ffa`5f802c03 48894d10        mov     qword ptr [rbp+10h],rcx
00007ffa`5f802c07 833d6a3f0c0000  cmp     dword ptr [00007ffa`5f8c6b78],0
00007ffa`5f802c0e 7405            je      00007ffa`5f802c15
00007ffa`5f802c10 e8fb9dcb5f      call    coreclr!JIT_DbgIsJustMyCode (00007ffa`bf4bca10)
00007ffa`5f802c15 48b9306d8c5ffa7f0000 mov rcx,7FFA5F8C6D30h
00007ffa`5f802c1f ba09000000      mov     edx,9
00007ffa`5f802c24 e8d783b95f      call    coreclr!JIT_GetSharedNonGCStaticBase_SingleAppDomain (00007ffa`bf39b000)
00007ffa`5f802c29 48b9982ef8d4fe010000 mov rcx,1FED4F82E98h
00007ffa`5f802c33 488b09          mov     rcx,qword ptr [rcx]
00007ffa`5f802c36 48894df8        mov     qword ptr [rbp-8],rcx

D:\MyCodes\DebugTest\DebugTest\DebugTest\Program.cs @ 5:
00007ffa`5f802c3a 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ffa`5f802c3e e8ed2ec95f      call    coreclr!GCInterface::GetGeneration (00007ffa`bf495b30)
00007ffa`5f802c43 8945f0          mov     dword ptr [rbp-10h],eax
00007ffa`5f802c46 8b4df0          mov     ecx,dword ptr [rbp-10h]
00007ffa`5f802c49 894df4          mov     dword ptr [rbp-0Ch],ecx

D:\MyCodes\DebugTest\DebugTest\DebugTest\Program.cs @ 6:
00007ffa`5f802c4c 8b4df4          mov     ecx,dword ptr [rbp-0Ch]
00007ffa`5f802c4f e85cffffff      call    00007ffa`5f802bb0
00007ffa`5f802c54 90              nop

D:\MyCodes\DebugTest\DebugTest\DebugTest\Program.cs @ 7:
00007ffa`5f802c55 e8fefbffff      call    00007ffa`5f802858 (System.Diagnostics.Debugger.Break(), mdToken: 00000000060063A7)
>>> 00007ffa`5f802c5a 90              nop

D:\MyCodes\DebugTest\DebugTest\DebugTest\Program.cs @ 8:
00007ffa`5f802c5b 488d4de0        lea     rcx,[rbp-20h]
00007ffa`5f802c5f e804fdffff      call    00007ffa`5f802968 (System.Console.ReadKey(), mdToken: 0000000006000032)
00007ffa`5f802c64 90              nop
00007ffa`5f802c65 90              nop
00007ffa`5f802c66 4883c440        add     rsp,40h
00007ffa`5f802c6a 5d              pop     rbp
00007ffa`5f802c6b c3              ret

```
```
0:000> !gcwhere 1FED4F82E98h
Address            Gen   Heap   segment            begin              allocated           size
000001FED4F82E98   4      0     000001FED4F80000   000001FED4F81000   000001FED4F85C38    0x0(0)
```

**Comment by svick** [comment]
I don't think objects directly referenced by `static` fields are special here: they start in generation 0 and increase their generation after every GC they survive, until they reach the maximum generation (2), like most other objects.

When I run this code:

```c#
var foo = Statics.Foo;
Console.WriteLine(GC.GetGeneration(foo));
GC.Collect();
Console.WriteLine(GC.GetGeneration(foo));
GC.Collect();
Console.WriteLine(GC.GetGeneration(foo));
```

[The output I get is:](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ6BuBDMACAMwHsi8BePAZWG2AEsBjAZwDoAxEgbnTiQE4AFAHEAwiyEBTYJKgSwtOkSgDiRAJRruaUSxFEANvokNgAzT37Cxk6RNnz6SlSQ1adew8dPm0vQTpsZOQUnVVd0HgQ8DiJ0AG90PDw4AGY8OihgPABJABMtJNTokgEMrLpctUS8BLQkpLzydPzqgF90doxfJAA2ZKjqBWZ46sK03j6Y4tIm2QB3ARSfVqA=)

```
0
1
2
```

On the other hand, string literals are special. When I run `GC.GetGeneration` on `Statics.str`, I always get 2.

Though there are some things I can't explain:

* Why  `GC.GetGeneration` returned 0 for you on `Statics.str`.
* Why does `!gcwhere` report generation 4.

**Comment by Li7ye** [comment]
@svick 
> * Why  `GC.GetGeneration` returned 0 for you on `Statics.str`.

I suppose you run the below code first in [sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ6BuBDMACAMwHsi8BePAZWG2AEsBjAZwDoAxEgbnTiQE4AFAHEAwiyEBTYJKgSwtOkSgDiRAJRruaUSxFEANvokNgAzT37Cxk6RNnz6SlSQ1adew8dPm0vQTpsZOQUnVVd0HgQ8DiJ0AG90PDw4AGY8OihgPABJABMtJNTokgEMrLpctUS8BLQkpLzydPzqgF90doxfJAA2ZKjqBWZ46sK03j6Y4tIm2QB3ARSfVqA=)
```csharp
var foo = Statics.Foo;
Console.WriteLine(GC.GetGeneration(foo));
GC.Collect();
Console.WriteLine(GC.GetGeneration(foo));
GC.Collect();
Console.WriteLine(GC.GetGeneration(foo));
```
Then you add `str` static string field into `Statics` and test it again. The output is always generation 2.

However, If you test `Statics.str` from scratch, the [output](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgAwFgBQiCM70DcCGYABAM7DEC8RAysPsAJYDGJAdGWANzpyYCcACgDiAYVZCApsElQJYegwD2UARwCUa7mlGsRigDb6JTYAM09+wsZOkTZ8xstXkNWnXsPHT5jJZ02ZOQUndR8eBCIAMUVFdABvdCIiOABmIgYoYCIASQATLSTUqJiBDKyGXLVEogS0JKS8oioKgqIAX3QOtAsANmSI2gUWeOrCtN4+3iRSciSqACJMBBT5rTagA=) is same as `Statics.Foo`.

**Comment by svick** [comment]
@Li7ye Interesting, so I was wrong about string literals. I guess SharpLab can reuse the process, which would explain why a string literal sometimes seems to start already at generation 2.

**Comment by cshung** [comment]
String literals are interesting. They have these properties:

1. Immutable,
2. Have no references,
3. Tends to live very long in the process (majority of them simply NEVER dies)
4. Have a big pile of them.
5. A good deal of them are NEVER accessed (e.g. exception messages)

These properties make them a great optimization target, and we attempted to, for example, https://github.com/dotnet/runtime/pull/49576.

How these binaries are compiled/published also impact them as well, for example, ready-to-run allocate them upfront during load, and nativeaot use frozen segments for them.

In short, it is moving target, we don't have a "correct" generation for them. It depends (on the compiler/runtime/mode/...) and it will change.

On the other hand, I will take a look at why `GC.GetGeneration` and `!gcwhere` disagrees. Feels like a bug there.

**Comment by Maoni0** [comment]
(I was on vacation and just saw this)

I dunno if you are doing anything special in your repro. I just created a .net 6 project and did this -

```csharp
using System;

namespace testStatics
{
    static class Statics
    {
        public static string str = "123";
    }
    class Program
    {
        static void Main(string[] args)
        {
            Console.ReadKey();
            var foo = Statics.str;
            var generation = GC.GetGeneration(foo);
            Console.WriteLine(generation);
        }
    }
}
```
I had it stop at ReadKey and attached windbg to it, in windbg I set a bp on `coreclr!GCInterface::GetGeneration` since that's what `GC.GetGeneration` calls, then I get the object address passed in and do a !gcwhere on it.

```
0:007> bp coreclr!GCInterface::GetGeneration
0:007> g
ModLoad: 000001fb`eeb70000 000001fb`eeb78000   C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.5\System.Text.Encoding.Extensions.dll
Breakpoint 0 hit
coreclr!GCInterface::GetGeneration:
00007ff9`52c95d60 4053            push    rbx
0:000> rrcx
rcx=000001fb8000c198
0:000> !do 000001fb8000c198
Name:        System.String
MethodTable: 00007ff8f309d698
EEClass:     00007ff8f3079dd8
Tracked Type: false
Size:        28(0x1c) bytes
File:        C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.5\System.Private.CoreLib.dll
String:      123
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007ff8f3089480  40002f2        8         System.Int32  1 instance                3 _stringLength
00007ff8f2feea10  40002f3        c          System.Char  1 instance               31 _firstChar
00007ff8f309d698  40002f1       e8        System.String  0   static 000001fb800013a0 Empty
0:000> !gcwhere 000001fb8000c198
Address            Gen   Heap   segment            begin              allocated           size
000001FB8000C198   0      0     000001FB80000000   000001FB80001000   000001FB8000D388    0x20(32)
```

both GetGeneration and !gcwhere return 0. is this different from what you are doing?
